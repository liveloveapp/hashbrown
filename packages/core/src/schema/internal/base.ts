/* eslint-disable @typescript-eslint/no-empty-object-type */
/* eslint-disable @typescript-eslint/no-explicit-any */
import {
  CleanInterfaceShape,
  Flatten,
  IsUnion,
  UnionToTuple,
} from '../../utils/types';

export const internal = '~schema';
export type internal = typeof internal;

type TypeInternals = {
  definition: {
    description: string;
    streaming: boolean;
  };
};

type TypeBox = {
  [internal]: TypeInternals;
};

export interface HashbrownTypeCtor<
  T extends TypeBox,
  D = T[internal]['definition'],
> {
  new (def: D): T;
  init(inst: T, def: D): asserts inst is T;
}

export const HashbrownTypeCtor = <
  T extends TypeBox,
  D extends TypeInternals['definition'] = T[internal]['definition'],
>(
  name: string,
  initializer: (instance: T, definition: D) => void,
): HashbrownTypeCtor<T, D> => {
  class Class {
    constructor(definition: D) {
      Class.init(this as any, definition);
    }

    static init(instance: T, definition: D) {
      instance[internal] ??= {
        definition: {
          description: '',
          streaming: false,
        },
      } as any;

      initializer(instance, definition);

      instance[internal].definition = definition;
    }
  }

  Object.defineProperty(Class, 'name', { value: name });

  return Class as HashbrownTypeCtor<T, D>;
};

interface HashbrownTypeDefinition {
  type:
    | 'string'
    | 'const-string'
    | 'number'
    | 'boolean'
    | 'integer'
    | 'object'
    | 'array'
    | 'enum'
    | 'any-of'
    | 'null';
  description: string;
  streaming: boolean;
}
export interface HashbrownType<out Result = unknown> {
  [internal]: HashbrownTypeInternals<Result>;
}

interface HashbrownTypeInternals<out Result = unknown>
  extends HashbrownType<Result> {
  definition: HashbrownTypeDefinition;
  result: Result;
}

export const HashbrownType: HashbrownTypeCtor<HashbrownType> =
  HashbrownTypeCtor('HashbrownType', (inst, def) => {
    inst ??= {} as any;
    inst[internal].definition = def;
  });

/**
 * --------------------------------------
 * --------------------------------------
 *             String Type
 * --------------------------------------
 * --------------------------------------
 */

interface StringTypeDefinition extends HashbrownTypeDefinition {
  type: 'string';
}

interface StringTypeInternals extends HashbrownTypeInternals<string> {
  definition: StringTypeDefinition;
}

export interface StringType extends HashbrownType<string> {
  [internal]: StringTypeInternals;
}

export const StringType: HashbrownTypeCtor<StringType> = HashbrownTypeCtor(
  'String',
  (inst, def) => {
    HashbrownType.init(inst, def);
  },
);

export function isStringType(type: HashbrownType): type is StringType {
  return type[internal].definition.type === 'string';
}

export function string(description: string): StringType {
  return new StringType({ type: 'string', description, streaming: false });
}

/**
 * --------------------------------------
 * --------------------------------------
 *          Const String Type
 * --------------------------------------
 * --------------------------------------
 */

interface ConstStringTypeDefinition<T extends string = string>
  extends HashbrownTypeDefinition {
  type: 'const-string';
  value: T;
}

interface ConstStringTypeInternals<T extends string = string>
  extends HashbrownTypeInternals<T> {
  definition: ConstStringTypeDefinition<T>;
}

export interface ConstStringType<T extends string = string>
  extends HashbrownType<T> {
  [internal]: ConstStringTypeInternals<T>;
}

export const ConstStringType: HashbrownTypeCtor<ConstStringType> =
  HashbrownTypeCtor('ConstString', (inst, def) => {
    HashbrownType.init(inst, def);
  });

export function isConstStringType(
  type: HashbrownType,
): type is ConstStringType {
  return type[internal].definition.type === 'const-string';
}

export function constString<T extends string>(value: T): ConstStringType<T> {
  return new ConstStringType({
    type: 'const-string',
    description: `${value}`,
    value,
    streaming: false,
  }) as any;
}
/**
 * --------------------------------------
 * --------------------------------------
 *             Number Type
 * --------------------------------------
 * --------------------------------------
 */

interface NumberTypeDefinition extends HashbrownTypeDefinition {
  type: 'number';
}

interface NumberTypeInternals extends HashbrownTypeInternals<number> {
  definition: NumberTypeDefinition;
}

export interface NumberType extends HashbrownType<number> {
  [internal]: NumberTypeInternals;
}

export const NumberType: HashbrownTypeCtor<NumberType> = HashbrownTypeCtor(
  'Number',
  (inst, def) => {
    HashbrownType.init(inst, def);
  },
);

export function isNumberType(type: HashbrownType): type is NumberType {
  return type[internal].definition.type === 'number';
}

export function number(description: string) {
  return new NumberType({ type: 'number', description, streaming: false });
}

/**
 * --------------------------------------
 * --------------------------------------
 *             Boolean Type
 * --------------------------------------
 * --------------------------------------
 */

interface BooleanTypeDefinition extends HashbrownTypeDefinition {
  type: 'boolean';
}

interface BooleanTypeInternals extends HashbrownTypeInternals<boolean> {
  definition: BooleanTypeDefinition;
}

export interface BooleanType extends HashbrownType<boolean> {
  [internal]: BooleanTypeInternals;
}

export const BooleanType: HashbrownTypeCtor<BooleanType> = HashbrownTypeCtor(
  'Boolean',
  (inst, def) => {
    HashbrownType.init(inst, def);
  },
);

export function isBooleanType(type: HashbrownType): type is BooleanType {
  return type[internal].definition.type === 'boolean';
}

export function boolean(description: string) {
  return new BooleanType({ type: 'boolean', description, streaming: false });
}

/**
 * --------------------------------------
 * --------------------------------------
 *             Integer Type
 * --------------------------------------
 * --------------------------------------
 */

interface IntegerTypeDefinition extends HashbrownTypeDefinition {
  type: 'integer';
}

interface IntegerTypeInternals extends HashbrownTypeInternals<number> {
  definition: IntegerTypeDefinition;
}

export interface IntegerType extends HashbrownType<number> {
  [internal]: IntegerTypeInternals;
}

export const IntegerType: HashbrownTypeCtor<IntegerType> = HashbrownTypeCtor(
  'Integer',
  (inst, def) => {
    HashbrownType.init(inst, def);
  },
);

export function isIntegerType(type: HashbrownType): type is IntegerType {
  return type[internal].definition.type === 'integer';
}

export function integer(description: string) {
  return new IntegerType({ type: 'integer', description, streaming: false });
}

/**
 * --------------------------------------
 * --------------------------------------
 *             Object Type
 * --------------------------------------
 * --------------------------------------
 */

type ObjectTypeResult<T extends Record<string, any>> = string extends keyof T
  ? object
  : {} extends T
    ? object
    : Flatten<{
        -readonly [K in keyof T]: T[K][internal]['result'];
      }>;

interface ObjectTypeDefinition<
  out Shape extends Record<string, any> = Record<string, any>,
> extends HashbrownTypeDefinition {
  type: 'object';
  readonly shape: Shape;
}

interface ObjectTypeInternals<Result extends Readonly<Record<string, any>>>
  extends HashbrownTypeInternals<ObjectTypeResult<Result>> {
  definition: ObjectTypeDefinition<Result>;
}

export interface ObjectType<
  Result extends Readonly<Record<string, any>> = Readonly<Record<string, any>>,
> extends HashbrownType {
  [internal]: ObjectTypeInternals<Result>;
}

export const ObjectType: HashbrownTypeCtor<ObjectType> = HashbrownTypeCtor(
  'Object',
  (inst, def) => {
    HashbrownType.init(inst, def);
  },
);

export function isObjectType(type: HashbrownType): type is ObjectType {
  return type[internal].definition.type === 'object';
}

export function object<Shape extends Record<string, any>>(
  description: string,
  shape: Shape,
): ObjectType<CleanInterfaceShape<Shape>> {
  return new ObjectType({
    type: 'object',
    description,
    streaming: false,
    shape,
  }) as any;
}

/**
 * --------------------------------------
 * --------------------------------------
 *             Array Type
 * --------------------------------------
 * --------------------------------------
 */

interface ArrayTypeDefinition<out Item extends HashbrownType = HashbrownType>
  extends HashbrownTypeDefinition {
  type: 'array';
  element: Item;
}

interface ArrayTypeInternals<Item extends HashbrownType = HashbrownType>
  extends HashbrownTypeInternals<Item[internal]['result'][]> {
  definition: ArrayTypeDefinition<Item>;
}

export interface ArrayType<Item extends HashbrownType = HashbrownType>
  extends HashbrownType {
  [internal]: ArrayTypeInternals<Item>;
}

export const ArrayType: HashbrownTypeCtor<ArrayType> = HashbrownTypeCtor(
  'Array',
  (inst, def) => {
    HashbrownType.init(inst, def);
  },
);

export function isArrayType(type: HashbrownType): type is ArrayType {
  return type[internal].definition.type === 'array';
}

export function array<Item extends HashbrownType>(
  description: string,
  item: Item,
): ArrayType<Item> {
  return new ArrayType({
    type: 'array',
    description,
    streaming: false,
    element: item,
  }) as any;
}

/**
 * --------------------------------------
 * --------------------------------------
 *             Any-Of Type
 * --------------------------------------
 * --------------------------------------
 */

interface AnyOfTypeDefinition<
  Options extends readonly HashbrownType[] = readonly HashbrownType[],
> extends HashbrownTypeDefinition {
  type: 'any-of';
  options: Options;
}

interface AnyOfTypeInternals<Options extends readonly HashbrownType[]>
  extends HashbrownTypeInternals<Options[number][internal]['result']> {
  definition: AnyOfTypeDefinition<Options>;
}

export interface AnyOfType<
  Options extends readonly HashbrownType[] = readonly HashbrownType[],
> extends HashbrownType<Options[number][internal]['result']> {
  [internal]: AnyOfTypeInternals<Options>;
}

export const AnyOfType: HashbrownTypeCtor<AnyOfType> = HashbrownTypeCtor(
  'AnyOfType',
  (inst, def) => {
    HashbrownType.init(inst, def);
  },
);

export function isAnyOfType(type: HashbrownType): type is AnyOfType {
  return type[internal].definition.type === 'any-of';
}

export function anyOf<const Options extends readonly HashbrownType[]>(
  description: string,
  options: Options,
): AnyOfType<Options> {
  return new AnyOfType({
    type: 'any-of',
    description,
    options,
    streaming: false,
  }) as any;
}

/**
 * --------------------------------------
 * --------------------------------------
 *             Enum Type
 * --------------------------------------
 * --------------------------------------
 */

interface EnumTypeDefinition<out Entries extends string[]>
  extends HashbrownTypeDefinition {
  type: 'enum';
  entries: Entries;
}

interface EnumTypeInternals<out Entries extends string[]>
  extends HashbrownTypeInternals<Entries> {
  definition: EnumTypeDefinition<Entries>;
}

export interface EnumType<out Entries extends string[] = string[]>
  extends HashbrownType<Entries> {
  [internal]: EnumTypeInternals<Entries>;
}

export const EnumType: HashbrownTypeCtor<EnumType> = HashbrownTypeCtor(
  'Enum',
  (inst, def) => {
    HashbrownType.init(inst, def);
  },
);

export function isEnumType(type: HashbrownType): type is EnumType {
  return type[internal].definition.type === 'enum';
}

export function enumType<Entries extends string[]>(
  description: string,
  entries: Entries,
) {
  return new EnumType({ type: 'enum', description, entries, streaming: false });
}

/**
 * --------------------------------------
 * --------------------------------------
 *             Null Type
 * --------------------------------------
 * --------------------------------------
 */

interface NullTypeDefinition extends HashbrownTypeDefinition {
  type: 'null';
}

interface NullTypeInternals extends HashbrownTypeInternals<null> {
  definition: NullTypeDefinition;
}

export interface NullType extends HashbrownType<null> {
  [internal]: NullTypeInternals;
}

export const NullType: HashbrownTypeCtor<NullType> = HashbrownTypeCtor(
  'Null',
  (inst, def) => {
    HashbrownType.init(inst, def);
  },
);

export function isNullType(type: HashbrownType): type is NullType {
  return type[internal].definition.type === 'null';
}

export function nullType(): NullType {
  return new NullType({ type: 'null', description: '', streaming: false });
}

/**
 * --------------------------------------
 * --------------------------------------
 *             Type Utilities
 * --------------------------------------
 * --------------------------------------
 */

export type Infer<T extends HashbrownType> = T[internal]['result'];

type SchemaForUnion<T> = AnyOfType<
  UnionToTuple<T> extends infer U
    ? U extends any[]
      ? { [K in keyof U]: Schema<U[K]> }
      : never
    : never
>;

export type Schema<T> =
  IsUnion<T> extends true
    ? SchemaForUnion<T>
    : T extends string[]
      ? EnumType<T>
      : T extends Array<infer U>
        ? ArrayType<Schema<U>>
        : T extends string
          ? string extends T
            ? StringType
            : ConstStringType<T>
          : T extends number
            ? NumberType | IntegerType
            : T extends boolean
              ? BooleanType
              : T extends null
                ? NullType
                : T extends object
                  ? ObjectType<{ [K in keyof T]: Schema<T[K]> }>
                  : never;
