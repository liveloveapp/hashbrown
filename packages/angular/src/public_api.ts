export { RenderMessageComponent } from './components';
export {
  chatResource,
  ChatResourceOptions,
  ChatResourceRef,
  completionResource,
  CompletionResourceOptions,
  CompletionResourceRef,
  structuredChatResource,
  StructuredChatResourceOptions,
  StructuredChatResourceRef,
  structuredCompletionResource,
  StructuredCompletionResourceOptions,
  StructuredCompletionResourceRef,
  uiChatResource,
  UiChatResourceOptions,
  UiChatResourceRef,
} from './resources';
export { provideHashbrown, ProvideHashbrownOptions } from './providers';
export {
  createTool,
  CreateToolInput,
  createToolWithArgs,
  CreateToolWithArgsInput,
  UiChatMessage,
  UiChatSchema,
  UiChatSchemaComponent,
  UiUserMessage,
  UiAssistantMessage,
  UiErrorMessage,
  exposeComponent,
  ComponentPropSchema,
  ExposedComponent,
} from './utils';
