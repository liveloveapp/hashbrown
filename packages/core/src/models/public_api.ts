export * as Api from './api.models';
export * as Internal from './internal.models';
export * from './view.models';
export * as helpers from './internal_helpers';
