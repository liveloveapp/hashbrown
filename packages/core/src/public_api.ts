export { fryHashbrown, type Hashbrown } from './hashbrown';
export * from './models';
export * from './schema';
export * as θcomponents from './utils/expose-component';
export * as θtypes from './utils/types';
export { type VendorClient } from './vendor/vendor-client';
