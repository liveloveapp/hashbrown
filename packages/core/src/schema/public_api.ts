export {
  anyOf,
  AnyOfType,
  array,
  ArrayType,
  boolean,
  BooleanType,
  literal as literal,
  LiteralType as ConstStringType,
  enumeration as enumeration,
  EnumType,
  HashbrownType,
  integer,
  IntegerType,
  isAnyOfType,
  isArrayType,
  isBooleanType,
  isLiteralType,
  isEnumType,
  isIntegerType,
  isNullType,
  isNumberType,
  isObjectType,
  isStringType,
  isHashbrownType,
  nullish,
  NullType,
  number,
  NumberType,
  object,
  ObjectType,
  string,
  StringType,
  type Infer,
  type Schema,
} from './base';
export { getDescription } from './get-description';
export * as streaming from './streaming';
export { toJsonSchema } from './to-json-schema';
