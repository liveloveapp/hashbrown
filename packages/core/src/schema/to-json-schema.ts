/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable @typescript-eslint/no-explicit-any */
import * as s from './base';
import { HashbrownType, internal } from './base';

/**
 * Convert an arbitrary description into a camelCase identifier.
 *
 * - Strips out any non-alphanumeric characters
 * - Splits on whitespace
 * - Lowercases all words, then uppercases the first letter of each subsequent word
 * - Prefixes with `_` if the result would start with a digit
 */
export function descriptionToCamelCase(description: string): string {
  const cleaned = description.replace(/[^a-zA-Z0-9]+/g, ' ').trim();
  const words = cleaned.split(/\s+/).map((w) => w.toLowerCase());
  if (words.length === 0) return '';
  const [first, ...rest] = words;
  const core =
    first + rest.map((w) => w[0].toUpperCase() + w.slice(1)).join('');
  return /^\d/.test(core) ? `_${core}` : core;
}

/**
 * Walks the HashbrownType graph, finds any sub-schemas seen more than once
 * (excluding the root), assigns each a unique name, and emits a draft-07 JSON Schema
 * with a $defs section.  Cycles always become $refs.
 */
export function toJsonSchema(schema: HashbrownType) {
  const rootNode = schema;

  // 1) Collect repeated nodes
  const seen = new Set<HashbrownType>();
  const repeats = new Set<HashbrownType>();
  (function visit(n: HashbrownType, path: HashbrownType[]) {
    if (seen.has(n)) {
      repeats.add(n);
      return;
    }
    seen.add(n);

    if (s.isObjectType(n)) {
      Object.values(n[internal].definition.shape).forEach((child) =>
        visit(child, [...path, n]),
      );
    } else if (s.isArrayType(n)) {
      visit(n[internal].definition.element, [...path, n]);
    } else if (s.isAnyOfType(n)) {
      n[internal].definition.options.forEach((opt) => visit(opt, [...path, n]));
    }
  })(rootNode, []);

  // never put the root itself into $defs
  repeats.delete(rootNode);

  // 2) Assign each repeated node a unique camelCase name
  const defNameMap = new Map<HashbrownType, string>();
  const usedNames = new Set<string>();
  let anon = 1;
  for (const node of repeats) {
    const desc = node[internal].definition.description || `def${anon++}`;
    let name = descriptionToCamelCase(desc) || `def${anon++}`;
    if (usedNames.has(name)) {
      let i = 1;
      while (usedNames.has(`${name}${i}`)) i++;
      name = `${name}${i}`;
    }
    usedNames.add(name);
    defNameMap.set(node, name);
  }

  /**
   * Recursive printer.
   *
   * @param n         current node
   * @param isRoot    true only for the very top-level schema
   * @param inDef     if non-null, we're printing $defs[inDef] — any other def becomes $ref
   * @param pathSeen  tracks the chain of inlined nodes to catch cycles
   */
  function printNode(
    n: HashbrownType,
    isRoot = false,
    inDef: HashbrownType | null = null,
    pathSeen: Set<HashbrownType> = new Set(),
  ): any {
    // a) cycle back to the root
    if (!isRoot && n === rootNode) {
      return { $ref: '#' };
    }

    // b) any other shared def becomes a $ref
    if (defNameMap.has(n) && n !== inDef) {
      const nm = defNameMap.get(n)!;
      return { $ref: `#/$defs/${nm}` };
    }

    // c) catch self-cycles or mutual cycles in inline portions
    if (pathSeen.has(n)) {
      // if it’s named, ref it; otherwise point at root
      if (defNameMap.has(n)) {
        const nm = defNameMap.get(n)!;
        return { $ref: `#/$defs/${nm}` };
      } else {
        return { $ref: '#' };
      }
    }

    // d) inline this node
    pathSeen.add(n);

    let result: any;

    if (s.isObjectType(n)) {
      // Sort props so that streaming ones are at the end
      const shapeWithStreamingAtEnd = Object.entries(
        n[internal].definition.shape,
      ).sort((a, b) => {
        if (!s.isStreaming(a[1]) && s.isStreaming(b[1])) {
          return -1;
        }
        if (s.isStreaming(a[1]) && !s.isStreaming(b[1])) {
          return 1;
        }

        return 0;
      });

      const props: Record<string, any> = {};
      for (const [k, child] of shapeWithStreamingAtEnd) {
        props[k] = printNode(child, false, inDef, pathSeen);
      }

      result = n.toJsonSchema();
      result.properties = props;
    } else if (s.isArrayType(n)) {
      result = n.toJsonSchema();
      result.items = printNode(
        n[internal].definition.element,
        false,
        inDef,
        pathSeen,
      );

      if (isRoot) {
        result = {
          type: 'object',
          additionalProperties: false,
          required: [s.PRIMITIVE_WRAPPER_FIELD_NAME],
          properties: {
            [s.PRIMITIVE_WRAPPER_FIELD_NAME]: result,
          },
        };
      }
    } else if (s.isAnyOfType(n)) {
      result = n.toJsonSchema();
      result.anyOf = n[internal].definition.options.map((opt, index) => {
        if (s.needsDiscriminatorWrapperInAnyOf(opt)) {
          const indexAsStr = `${index}`;
          return {
            type: 'object',
            additionalProperties: false,
            required: [indexAsStr],
            properties: {
              [indexAsStr]: printNode(opt, false, inDef, pathSeen),
            },
          };
        } else {
          return printNode(opt, false, inDef, pathSeen);
        }
      });
    } else {
      result = n.toJsonSchema();

      if (isRoot) {
        result = {
          type: 'object',
          additionalProperties: false,
          required: [s.PRIMITIVE_WRAPPER_FIELD_NAME],
          properties: {
            [s.PRIMITIVE_WRAPPER_FIELD_NAME]: result,
          },
        };
      }
    }

    pathSeen.delete(n);
    return result;
  }

  // 3) Build the $defs section
  const defs: Record<string, any> = {};
  for (const [node, name] of defNameMap.entries()) {
    // isRoot=false, inDef=node, fresh pathSeen
    defs[name] = printNode(node, false, node, new Set());
  }

  // 4) Print the root schema
  const rootPrinted = printNode(rootNode, true, null, new Set());

  // 5) Assemble and return
  return {
    $schema: 'http://json-schema.org/draft-07/schema#',
    ...rootPrinted,
    ...(Object.keys(defs).length > 0 ? { $defs: defs } : {}),
  };
}
