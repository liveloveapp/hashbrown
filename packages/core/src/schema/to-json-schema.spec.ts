import * as s from './public_api';
import { PRIMITIVE_WRAPPER_FIELD_NAME } from './base';
import { descriptionToCamelCase, toJsonSchema } from './to-json-schema';

test('descriptionToCamelCase variants snapshot', () => {
  const samples = [
    'Hello world',
    '  Title: With Punctuations!!  ',
    '123 starts with number',
    'alreadyCamel',
    'kebab-case and snake_case',
    'multi    space\t and\nlines',
    'MiXeD CaSe Words',
  ];

  const result = Object.fromEntries(
    samples.map((x) => [x, descriptionToCamelCase(x)]),
  );

  expect(result).toMatchSnapshot();
});

test('root: wraps primitive string under __wrappedPrimitive', () => {
  const schema = s.string('A string');
  const json = toJsonSchema(schema);
  expect(json).toMatchSnapshot();
  expect(json.properties[PRIMITIVE_WRAPPER_FIELD_NAME]).toBeDefined();
});

test('root: wraps array under __wrappedPrimitive', () => {
  const schema = s.array('Arr', s.number('num'));
  const json = toJsonSchema(schema);
  expect(json).toMatchSnapshot();
});

test('object: properties, required, additionalProperties=false', () => {
  const schema = s.object('Root obj', {
    a: s.number('A'),
    b: s.string('B'),
  });
  const json = toJsonSchema(schema);
  expect(json).toMatchSnapshot();
});

test('defs: repeated sub-schema is extracted into $defs and referenced', () => {
  const shared = s.object('Shared Item', { x: s.number('x') });
  const schema = s.object('Root', { a: shared, b: shared });
  const json = toJsonSchema(schema);
  expect(json).toMatchSnapshot();
});

test('anyOf: primitives are direct without wrappers', () => {
  const schema = s.anyOf([s.number('N'), s.string('S')]);
  const json = toJsonSchema(schema);
  expect(json).toMatchSnapshot();
});

test('anyOf: object options default to index wrappers', () => {
  const schema = s.anyOf([
    s.object('A', { a: s.number('a') }),
    s.object('B', { b: s.string('b') }),
  ]);
  const json = toJsonSchema(schema);
  expect(json).toMatchSnapshot();
});

test('anyOf: mixed options use wrappers only for complex ones', () => {
  const schema = s.anyOf([
    s.string('S'),
    s.object('A', { a: s.number('a') }),
    s.number('N'),
  ]);
  const json = toJsonSchema(schema);
  expect(json).toMatchSnapshot();
});

test('anyOf: single unique string-literal per object uses literal envelopes, omitting discriminator inside', () => {
  const schema = s.anyOf([
    s.object('Show markdown', {
      $tagName: s.literal('app-markdown'),
      $props: s.object('Props', { data: s.string('data') }),
    }),
    s.object('Show button', {
      $tagName: s.literal('app-button'),
      $props: s.object('Props', { data: s.string('data') }),
    }),
  ]);
  const json = toJsonSchema(schema);
  expect(json).toMatchSnapshot();
  const markdownWrapper = json.anyOf.find((x: any) =>
    Object.prototype.hasOwnProperty.call(x.properties, 'app-markdown'),
  );
  expect(
    markdownWrapper.properties['app-markdown'].properties['$tagName'],
  ).toBeUndefined();
});

test('anyOf: duplicate literal values fall back to index wrappers', () => {
  const schema = s.anyOf([
    s.object('A', { type: s.literal('dup'), a: s.number('a') }),
    s.object('B', { type: s.literal('dup'), b: s.string('b') }),
  ]);
  const json = toJsonSchema(schema);
  expect(json).toMatchSnapshot();
});

test('anyOf: multiple literals per object fall back to index wrappers', () => {
  const schema = s.anyOf([
    s.object('A', { t1: s.literal('x'), t2: s.literal('y'), a: s.number('a') }),
    s.object('B', { t1: s.literal('p'), t2: s.literal('q'), b: s.string('b') }),
  ]);
  const json = toJsonSchema(schema);
  expect(json).toMatchSnapshot();
});
