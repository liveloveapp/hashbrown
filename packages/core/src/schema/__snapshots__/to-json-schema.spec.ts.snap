// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`anyOf: duplicate literal values fall back to index wrappers 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "anyOf": [
    {
      "additionalProperties": false,
      "properties": {
        "0": {
          "additionalProperties": false,
          "description": "A",
          "properties": {
            "a": {
              "description": "a",
              "type": "number",
            },
            "type": {
              "const": "dup",
              "description": "dup",
              "type": "string",
            },
          },
          "required": [
            "type",
            "a",
          ],
          "type": "object",
        },
      },
      "required": [
        "0",
      ],
      "type": "object",
    },
    {
      "additionalProperties": false,
      "properties": {
        "1": {
          "additionalProperties": false,
          "description": "B",
          "properties": {
            "b": {
              "description": "b",
              "type": "string",
            },
            "type": {
              "const": "dup",
              "description": "dup",
              "type": "string",
            },
          },
          "required": [
            "type",
            "b",
          ],
          "type": "object",
        },
      },
      "required": [
        "1",
      ],
      "type": "object",
    },
  ],
}
`;

exports[`anyOf: mixed options use wrappers only for complex ones 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "anyOf": [
    {
      "description": "S",
      "type": "string",
    },
    {
      "additionalProperties": false,
      "properties": {
        "1": {
          "additionalProperties": false,
          "description": "A",
          "properties": {
            "a": {
              "description": "a",
              "type": "number",
            },
          },
          "required": [
            "a",
          ],
          "type": "object",
        },
      },
      "required": [
        "1",
      ],
      "type": "object",
    },
    {
      "description": "N",
      "type": "number",
    },
  ],
}
`;

exports[`anyOf: multiple literals per object fall back to index wrappers 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "anyOf": [
    {
      "additionalProperties": false,
      "properties": {
        "0": {
          "additionalProperties": false,
          "description": "A",
          "properties": {
            "a": {
              "description": "a",
              "type": "number",
            },
            "t1": {
              "const": "x",
              "description": "x",
              "type": "string",
            },
            "t2": {
              "const": "y",
              "description": "y",
              "type": "string",
            },
          },
          "required": [
            "t1",
            "t2",
            "a",
          ],
          "type": "object",
        },
      },
      "required": [
        "0",
      ],
      "type": "object",
    },
    {
      "additionalProperties": false,
      "properties": {
        "1": {
          "additionalProperties": false,
          "description": "B",
          "properties": {
            "b": {
              "description": "b",
              "type": "string",
            },
            "t1": {
              "const": "p",
              "description": "p",
              "type": "string",
            },
            "t2": {
              "const": "q",
              "description": "q",
              "type": "string",
            },
          },
          "required": [
            "t1",
            "t2",
            "b",
          ],
          "type": "object",
        },
      },
      "required": [
        "1",
      ],
      "type": "object",
    },
  ],
}
`;

exports[`anyOf: object options default to index wrappers 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "anyOf": [
    {
      "additionalProperties": false,
      "properties": {
        "0": {
          "additionalProperties": false,
          "description": "A",
          "properties": {
            "a": {
              "description": "a",
              "type": "number",
            },
          },
          "required": [
            "a",
          ],
          "type": "object",
        },
      },
      "required": [
        "0",
      ],
      "type": "object",
    },
    {
      "additionalProperties": false,
      "properties": {
        "1": {
          "additionalProperties": false,
          "description": "B",
          "properties": {
            "b": {
              "description": "b",
              "type": "string",
            },
          },
          "required": [
            "b",
          ],
          "type": "object",
        },
      },
      "required": [
        "1",
      ],
      "type": "object",
    },
  ],
}
`;

exports[`anyOf: primitives are direct without wrappers 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "anyOf": [
    {
      "description": "N",
      "type": "number",
    },
    {
      "description": "S",
      "type": "string",
    },
  ],
}
`;

exports[`anyOf: single unique string-literal per object uses literal envelopes, omitting discriminator inside 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "anyOf": [
    {
      "additionalProperties": false,
      "properties": {
        "app-markdown": {
          "additionalProperties": false,
          "description": "Show markdown",
          "properties": {
            "$props": {
              "additionalProperties": false,
              "description": "Props",
              "properties": {
                "data": {
                  "description": "data",
                  "type": "string",
                },
              },
              "required": [
                "data",
              ],
              "type": "object",
            },
          },
          "required": [
            "$props",
          ],
          "type": "object",
        },
      },
      "required": [
        "app-markdown",
      ],
      "type": "object",
    },
    {
      "additionalProperties": false,
      "properties": {
        "app-button": {
          "additionalProperties": false,
          "description": "Show button",
          "properties": {
            "$props": {
              "additionalProperties": false,
              "description": "Props",
              "properties": {
                "data": {
                  "description": "data",
                  "type": "string",
                },
              },
              "required": [
                "data",
              ],
              "type": "object",
            },
          },
          "required": [
            "$props",
          ],
          "type": "object",
        },
      },
      "required": [
        "app-button",
      ],
      "type": "object",
    },
  ],
}
`;

exports[`defs: repeated sub-schema is extracted into $defs and referenced 1`] = `
{
  "$defs": {
    "sharedItem": {
      "additionalProperties": false,
      "description": "Shared Item",
      "properties": {
        "x": {
          "description": "x",
          "type": "number",
        },
      },
      "required": [
        "x",
      ],
      "type": "object",
    },
  },
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "description": "Root",
  "properties": {
    "a": {
      "$ref": "#/$defs/sharedItem",
    },
    "b": {
      "$ref": "#/$defs/sharedItem",
    },
  },
  "required": [
    "a",
    "b",
  ],
  "type": "object",
}
`;

exports[`descriptionToCamelCase variants snapshot 1`] = `
{
  "  Title: With Punctuations!!  ": "titleWithPunctuations",
  "123 starts with number": "_123StartsWithNumber",
  "Hello world": "helloWorld",
  "MiXeD CaSe Words": "mixedCaseWords",
  "alreadyCamel": "alreadycamel",
  "kebab-case and snake_case": "kebabCaseAndSnakeCase",
  "multi    space	 and
lines": "multiSpaceAndLines",
}
`;

exports[`object: properties, required, additionalProperties=false 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "description": "Root obj",
  "properties": {
    "a": {
      "description": "A",
      "type": "number",
    },
    "b": {
      "description": "B",
      "type": "string",
    },
  },
  "required": [
    "a",
    "b",
  ],
  "type": "object",
}
`;

exports[`root: wraps array under __wrappedPrimitive 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "__wrappedPrimitive": {
      "description": "Arr",
      "items": {
        "description": "num",
        "type": "number",
      },
      "type": "array",
    },
  },
  "required": [
    "__wrappedPrimitive",
  ],
  "type": "object",
}
`;

exports[`root: wraps primitive string under __wrappedPrimitive 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "__wrappedPrimitive": {
      "description": "A string",
      "type": "string",
    },
  },
  "required": [
    "__wrappedPrimitive",
  ],
  "type": "object",
}
`;
